File: typescript/src/db/pool.ts
import pg from "pg";

export function createPool(connectionString?: string): pg.Pool {
  return new pg.Pool({
    connectionString:
      connectionString ?? "postgresql://postgres:postgres@localhost:5433/interview",
  });
}

export type PoolOrClient = pg.Pool | pg.PoolClient;

export async function withTransaction<T>(
  pool: pg.Pool,
  fn: (client: pg.PoolClient) => Promise<T>,
): Promise<T> {
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    const result = await fn(client);
    await client.query("COMMIT");
    return result;
  } catch (err) {
    await client.query("ROLLBACK");
    throw err;
  } finally {
    client.release();
  }
}

File: python/src/db/pool.py
import os
from typing import Awaitable, Callable, TypeVar

import asyncpg

T = TypeVar("T")

DEFAULT_DSN = "postgresql://postgres:postgres@localhost:5433/interview"


async def create_pool(dsn: str | None = None) -> asyncpg.Pool:
    return await asyncpg.create_pool(dsn=dsn or os.environ.get("DATABASE_URL", DEFAULT_DSN))


async def with_transaction(
    pool: asyncpg.Pool,
    fn: Callable[[asyncpg.Connection], Awaitable[T]],
) -> T:
    async with pool.acquire() as conn:
        async with conn.transaction():
            return await fn(conn)
